### ISSUE TITLE
Asynchronous OpenCL calls have potential for undefined behaviour

### ISSUE THREAD

**Issue by t4c1** [stan-dev/math#1346]
## Description
There are some innocent-looking code patterns that can actually result in reading memory from destroyed objects. 

<s>1) Using any temporary `matrix_cl` variable in a function that calls a kernel. Synce kernel call is asynchrounous that temporary could be destroyed before the kernel is executed. This is only problem with arguments of type `matrix_cl` - `setKernelArguments()` makes a copy of scalars, but not of `cl::Buffer`s.</s>

2) Constructing a `matrix_cl` from any temporary object. Since we are using asynchronous copying that that temporary could be destroyed before copying happens.

## Example

<s>1)
```
N=100;
matrix_cl<double> a(N,N), b(N,N), c(N,N);
matrix_cl<double> d = a + b + c;
```
</s>

2)
```
MatrixXd a = ...
MatrixXd b = ...

matrix_cl<double> c(a + b);
```

## Expected Output
Well-defined behaviour.

#### Current Version:
v2.20.0

**Comment by t4c1** [comment]
I just found out that documentation for `clReleaseMemoryObject` says that a memory object is not deleted until all comands using it have completed, even if its reference count is 0. So 1 is not actually an issue.

**Comment by SteveBronder** [comment]
Ty! For (2) could we just make a copy of the object when it's a temp?

**Comment by t4c1** [comment]
I guess we could allocate the object on the heap and construct it by moving from the temporary. Than bind a callback to the event that will delete the object. That requires two versions of every existing `matrix_cl` constructor, but I think it is a good option.

**Comment by SteveBronder** [comment]
I think we could also use more generic templates and then for temporaries have an `if(is_rvalue_reference<T>::value) // do copy of temp`

**Comment by t4c1** [comment]
You are right. Two versions of constructors are not needed. However the copies still need to be dynamically allocated.

**Comment by SteveBronder** [comment]
Which devices and platforms did the tests fail on? Looking at your clinfo it looks like your drivers support out of order while the Intel(R) HD Graphics 4000 device does not. If I had to guess I bet we are only looking at if the driver supports it and not the device. I ran the test you posted on my laptop (which does not support async) and can confirm I'm getting it here as well. I'll work on a fix tmrw
